### DISCUSSION TITLE
Typescript typing of both <Head /> components, custom component children

### DISCUSSION THREAD

**Discussion by cyrilchapon** [vercel/next.js#64653] [Ideas]
# Bug report

## Describe the bug

After reading [#17721](https://github.com/vercel/next.js/issues/17721), [#6919](https://github.com/vercel/next.js/issues/6919), and [#13230](https://github.com/vercel/next.js/issues/13230) and the [docs here](https://nextjs.org/docs/api-reference/next/head), I just realized after several hours that my wrapping of `<link>`s in components inside the `<Head>` (**BOTH** the *document* one and the *app* one) was not working. This is pretty subtle so I'll try to explain the double-bug stuff.

> title, meta or any other elements (e.g. script) need to be contained as direct children of the Head element, or wrapped into maximum one level of <React.Fragment> or arrays—otherwise the tags won't be correctly picked up on client-side navigations.

This is a double bug that has the same root :

1. The Document's Head component (`import { Head } from 'next/document'`) **doesn't** show the nice warning
> Warning: viewport meta tags should not be used in _document.js's . https://err.sh/next.js/no-document-viewport-meta
when a custom component is wrongly passed to it. Furthermore; it **works** (renders correctly) with a custom-component in this case.
2. The Typescript type of App's Head component (`import Head from 'next/head'`) **accepts** a custom component instead of links and stuff. This leads to a runtime bug that could be avoided (I think with narrowing the `children` type)

## To Reproduce

- Create a custom component that wraps a link

```
const ViewportMetaLink = () => (
  <meta
    name='viewport'
    content='minimum-scale=1, initial-scale=1, width=device-width'
  />
)
```

- Try to use it in the Document's head

```
class Document extends _Document<DocumentInitialProps> {
  render() {
    return (
      <Html>
        <Head>
          <ViewportMetaLink />
        </Head>

        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}
```
- (Notice how this just works, without any warning. But has the duplicate-meta-bug #6919
- Remove the previous, and try it in the App's Head
```
const App = ({ Component, pageProps }: AppProps) => (
  <>
    <Head>
      <ViewportMetaLink />
    </Head>

    <div />
  </>
)
```
- (Notice how the Typescript types allow it; but lead to a runtime error #17721 )

## Expected behavior

I'd expect to have followed that path at least :
- Trying inside Document; having seen the warning
- Trying inside App; having been blocked by Typescript types (allowing only html elements, not custom components)

Even better :
- Trying inside Document; having blocked by Typescript types (not allowing any children at all, since it's discouraged ?)
- Trying inside App; having been blocked by Typescript types (allowing only html elements, not custom components)

## Screenshots

- Custom Components in Document Head children are OK for Typescript
<img width="461" alt="Capture d’écran 2020-11-14 à 12 38 03" src="https://user-images.githubusercontent.com/10728426/99146239-88ea8780-2676-11eb-84e4-e43586b8191a.png">

- And no warning is shown in console in that case
<img width="972" alt="Capture d’écran 2020-11-14 à 12 38 45" src="https://user-images.githubusercontent.com/10728426/99146247-9acc2a80-2676-11eb-9a79-cde92ecd23f7.png">

<img width="460" alt="Capture d’écran 2020-11-14 à 12 38 51" src="https://user-images.githubusercontent.com/10728426/99146256-a4ee2900-2676-11eb-8400-4c61efcd2bc4.png">

- Custom Components in App Head children are OK for Typescript
<img width="428" alt="Capture d’écran 2020-11-14 à 12 39 10" src="https://user-images.githubusercontent.com/10728426/99146265-bc2d1680-2676-11eb-9313-4901255179af.png">

- And they lead to the logical runtime issue
<img width="995" alt="Capture d’écran 2020-11-14 à 12 17 43" src="https://user-images.githubusercontent.com/10728426/99146270-cb13c900-2676-11eb-88c1-d4eb8c8d7c2b.png">

## System information

- OS: macOS
- Browser (if applies) All
- Next.js 10.0.0
- Node v12.19.0
- Typescript 4.0.5

## Additional context

tsconfig.json
```
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "rootDir": "./src",
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "baseUrl": "./src",
    "paths": {
      "$styles": ["styles"],
      "$styles/*": ["styles/*"],
      "$components": ["components"],
      "$components/*": ["components/*"],
      "$constants": ["constants"],
      "$constants/*": ["constants/*"],
    }
  },
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx"
  ],
  "exclude": [
    "node_modules"
  ]
}

```

**Comment by cyrilchapon** [comment]
😂 wtf

**Comment by cyrilchapon** [comment]
I think you should take a look at your [notification settings](https://docs.github.com/es/free-pro-team@latest/github/managing-subscriptions-and-notifications-on-github/configuring-notifications)

Cheers

**Comment by TyMick** [comment]
Hi, @Kevinfoss— Here are a few solutions you can try that should cancel all your GitHub email subscriptions:

- Go to https://github.com/watching
  1. Click the "Unwatch all" button
- Go to https://github.com/notifications/subscriptions
  1. Click the top check box next to "Subscriptions"
  2. Click the "Unsubscribe" button
- Go to https://github.com/settings/emails
  1. Scroll down to the bottom
  2. Click "Only receive account related emails, and those I subscribe to"
  3. Click the "Save email preferences" button
- Go to https://github.com/settings/notifications
  1. Uncheck any boxes that mention "email"

Alternatively, if you don't use GitHub anymore, you can always go to https://github.com/settings/admin, scroll down to the bottom, and click the "Delete your account" button.

**Comment by Adamih** [comment]
Can I work on this?

**Comment by Aribaskar-jb** [comment]
Can I work on this issue you can Assign me this task
